### PR TITLE
[jit] Add signature checking for CALLI.

### LINKED ISSUE #7792: Failure to AOT compile due to condition `load_opcode != OP_LOADV_MEMBASE' not met
_From @mellinoe on March 22, 2018 7:32_

### Steps to Reproduce

1. Create a Xamarin.iOS application.
2. Add a reference to this NuGet package: https://www.nuget.org/packages/Vk/
3. Try to build

### Expected Behavior
It works.

### Actual Behavior
An error is given and compilation fails.

https://github.com/mono/mono/blob/master/mono/mini/method-to-ir.c#L14027

```
Target _CompileToNative:
    Mono Ahead of Time compiler - compiling assembly /Users/eric/projects/veldrid-samples/bin/obj/TexturedCube.iOS/iPhone/Debug/device-builds/iphone7.2-11.2.6/mtouch-cache/Build/vk.dll
    AOTID 8AB2CCE5-F2F1-F5F5-4C24-C94BC0E363F2
    * Assertion at /Users/builder/data/lanes/5944/4d8c4e32/source/xamarin-macios/external/mono/mono/mini/method-to-ir.c:14029, condition `load_opcode != OP_LOADV_MEMBASE' not met
    
    
    MTOUCH : error MT3001: Could not AOT the assembly '/Users/eric/projects/veldrid-samples/bin/obj/TexturedCube.iOS/iPhone/Debug/device-builds/iphone7.2-11.2.6/mtouch-cache/Build/vk.dll'
```
### Environment

```
=== Visual Studio Community 2017 for Mac (Preview) ===

Version 7.5 Preview (7.5 build 933)
Installation UUID: b1e6ad1a-54b0-46de-8e03-9ed2a705273e
Runtime:
	Mono 5.10.0.179 (2017-12/dac60775616) (64-bit)
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 4.4.0.36 (master / 0c7c49a6)

	Package version: 510000179

=== NuGet ===

Version: 4.3.1.4445

=== .NET Core ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	2.1.0-preview1-26216-03
	2.0.5
	2.0.3
SDK: /usr/local/share/dotnet/sdk/2.1.300-preview1-008174/Sdks
SDK Versions:
	2.1.300-preview1-008174
	2.1.4
	2.1.2
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/5.10.0/lib/mono/msbuild/15.0/bin/Sdks

=== Xamarin.Profiler ===

Version: 1.6.2
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Xamarin.Android ===

Version: 8.3.0.5 (Visual Studio Community)
Android SDK: /Users/eric/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		7.1 (API level 25)

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 26.0.2
SDK Build Tools Version: 26.0.3

Java SDK: /usr
java version "1.8.0_131"
Java(TM) SE Runtime Environment (build 1.8.0_131-b11)
Java HotSpot(TM) 64-Bit Server VM (build 25.131-b11, mixed mode)

Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Xamarin Inspector ===

Version: 1.4.0
Hash: b3f92f9
Branch: master
Build date: Fri, 19 Jan 2018 22:00:34 GMT
Client compatibility: 1

=== Apple Developer Tools ===

Xcode 9.2 (13772)
Build 9C40b

=== Xamarin.Mac ===

Version: 4.4.0.31 (Visual Studio Community)

=== Xamarin.iOS ===

Version: 11.10.0.32 (Visual Studio Community)
Hash: 4d8c4e32
Branch: d15-7
Build date: 2018-03-07 18:39:46-0500

=== Build Information ===

Release ID: 705000933
Git revision: ace1e107c88477db1abe5e807809650f064112c4
Build date: 2018-03-19 17:42:41-04
Xamarin addins: a01263bdbba2589b1f6e42ef085bd31f6e3bfcf4
Build lane: monodevelop-lion-d15-7

=== Operating System ===

Mac OS X 10.13.3
Darwin 17.4.0 Darwin Kernel Version 17.4.0
    Sun Dec 17 09:19:54 PST 2017
    root:xnu-4570.41.2~1/RELEASE_X86_64 x86_64

=== Enabled user installed addins ===

Internet of Things (IoT) development (Preview) 7.1
```

_Copied from original issue: xamarin/xamarin-macios#3798_

### PR DESCRIPTION
**Pull request by vargaz**
Fixes #7792.

### CODE REVIEW

**Comment by marek-safar** [comment]
@monojenkins backport 2017-12

**Comment by marek-safar** [comment]
@monojenkins backport 2018-02

File: mono/mini/iltests.il [hunk 1]
```diff
@@ -3136,4 +3136,43 @@ L_3:
...
+	ret
+   }
+
+  .method public static int32 calli_sig_check_2 () cil managed
+  {
+	.maxstack 16
+
+	ldc.r4 1.13
+	ldftn int32 Tests::calli_sig_check(string)
+	calli int32(string)
```
[the remark this answers is not in the document]
Comment (vargaz) [reply mono/mini/iltests.il]: There is an r4 on the stack, not a string.

**Comment by vargaz** [comment]
The newly added test fails with the interpreter.

File: mono/mini/method-to-ir.c [hunk 2]
```diff
@@ -8237,6 +8237,9 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 
 			sp -= n;
 
+			if (!(cfg->method->wrapper_type && cfg->method->wrapper_type != MONO_WRAPPER_DYNAMIC_METHOD) && check_call_signature (cfg, fsig, sp))
```
Comment (jaykrell) [inline mono/mini/method-to-ir.c]: This wrapper check seems to recur. Can we make a function for it? I think tailcall checks it also.

**Comment by lewurm** [comment]
let's ignore it on the interpreter for now, we can fix it later

**Comment by marek-safar** [comment]
We need to exclude the test for the interpreter

File: mono/mini/method-to-ir.c (hunk 2 above)
Comment (kumpera) [reply mono/mini/method-to-ir.c]: Kinda odd we're doing it for dynamic methods here, this doesn't feel like the sort of constraint they need to violate -  it's usually just visibility.

File: mono/mini/iltests.il (hunk 1 above)
[the remark this answers is not in the document]
Comment (kumpera) [reply mono/mini/iltests.il]: oh! I though this was about the function pointer and the calli signature.

**Comment by marek-safar** [comment]
@vargaz could you add it to the interpreter tests exclusion list

**Comment by lambdageek** [comment]
@monojenkins build Linux x64 Checked Private Types Build

**Comment by vargaz** [comment]
@monojenkins backport 2018-02

**Comment by vargaz** [comment]
@monojenkins backport 2017-12

**Pull request merged**